File: fc-chat/demo/part3/ch4/message-system-add-push/src/main/java/net/prostars/messagesystem/auth/MessageUserDetails.java
package net.prostars.messagesystem.auth;

import com.fasterxml.jackson.annotation.*;
import java.util.Collection;
import java.util.List;
import java.util.Objects;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "@class")
@JsonIgnoreProperties(ignoreUnknown = true)
public class MessageUserDetails implements UserDetails {

  private final Long userId;
  private final String username;
  private String password;

  @JsonCreator
  public MessageUserDetails(
      @JsonProperty("userId") Long userId,
      @JsonProperty("username") String username,
      @JsonProperty("password") String password) {
    this.userId = userId;
    this.username = username;
    this.password = password;
  }

  public Long getUserId() {
    return userId;
  }
  
  public void erasePassword() {
    password = "";
  }

  @Override
  public String getUsername() {
    return username;
  }

  @Override
  public String getPassword() {
    return password;
  }

  @Override
  @JsonIgnore
  public Collection<? extends GrantedAuthority> getAuthorities() {
    return List.of();
  }

  @Override
  public boolean equals(Object o) {
    if (o == null || getClass() != o.getClass()) return false;
    MessageUserDetails that = (MessageUserDetails) o;
    return Objects.equals(username, that.username);
  }

  @Override
  public int hashCode() {
    return Objects.hashCode(username);
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/auth/MessageUserDetails.kt
package com.yscorp.withpush.messagesystem.auth

import com.fasterxml.jackson.annotation.*
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import java.util.*

@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "@class")
@JsonIgnoreProperties(ignoreUnknown = true)
class MessageUserDetails @JsonCreator constructor(
    @param:JsonProperty("userId") val userId: Long,
    @param:JsonProperty("username") private val username: String,
    @param:JsonProperty("password") private var password: String
) : UserDetails {
    fun erasePassword() {
        password = ""
    }

    override fun getUsername(): String {
        return username
    }

    override fun getPassword(): String {
        return password
    }

    @JsonIgnore
    override fun getAuthorities(): Collection<GrantedAuthority> {
        return listOf()
    }

    override fun equals(o: Any?): Boolean {
        if (o == null || javaClass != o.javaClass) return false
        val that = o as MessageUserDetails
        return username == that.username
    }

    override fun hashCode(): Int {
        return Objects.hashCode(username)
    }
}
